numpyro/infer/util.py

```diff
@@ -61,16 +61,20 @@ def log_density(model, model_args, model_kwargs, params, skip_dist_transforms=Fa
         if site['type'] == 'sample':
             value = site['value']
             intermediates = site['intermediates']
+            mask = site['mask']
+            scale = site['scale']
+            # exit early if mask all
+            if not np.any(mask):
+                return device_put(0.), model_trace
             if intermediates:
                 if skip_dist_transforms:
                     log_prob = site['fn'].base_dist.log_prob(intermediates[0][0])
                 else:
                     log_prob = site['fn'].log_prob(value, intermediates)
             else:
                 log_prob = site['fn'].log_prob(value)
+            log_prob = np.where(mask, scale * log_prob, 0.)
             log_prob = np.sum(log_prob)
-            if 'scale' in site:
```

Comment:
WDYT about this pattern? I guess it would be nice to have this check to save some computations, but I don't recall why we have `scale` (and `mask`) in `sample` primitive by default. :(

Reply:
I think it is okay to have some default value for `scale` or `mask`, and we can always add other optimizations like in Pyro where we check if the scale is identically zero or whether all values are masked out and treat these cases specially. I found no performance difference without these minor optimizations on the few examples I ran in NumPyro. e.g. in this case we only end up multiplying by a scalar 1 which probably makes no difference under JIT (in fact, we'll need to use `lax.cond` to apply these optimizations correctly under JIT).